### PR TITLE
World

### PR DESCRIPTION
**Pull request by criticalglitch**
This is the first step toward completing Issue#86; world generation functions have been pulled out into a new World class.

### CODE REVIEW

**Review by bakursait2 — CHANGES_REQUESTED** [review]
It is a perfect way to split the code as SOLID recommends. However, The test failed. Can you fix the related issue?

**Comment by criticalglitch** [comment]
> It is a perfect way to split the code as SOLID recommends. However, The test failed. Can you fix the related issue?

Turns out the python version in CI was a lower version than what poetry was specifying. I have pushed a fix to upgrade it and CI now seems happy.

**Review by bakursait2 — APPROVED** [review]

**Review by xieWright — COMMENTED** [review]
I feel like before I can approve this, I need to understand why we need to split the model into model and world. As far as I can tell, when you move everything to world, you're pretty much just making a collection of static functions that the model will call. However most of these functions were also important to the model function. What exactly is the difference between the model and the world?

File: tempus_fugit_minecraft/world.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+            center of the cloud.
+        @param cloud_center_x Represents the z-coordinate center of the 
+            cloud.
+        @param s represent number of blocks drawn from center - goes in 
+            each direction around the center.
+        
+        @return single_cloud A list that contains pairs of blocks and
+            positions that represent a cloud. 
+        @see [Issue#20](https://github.com/WSUCEG-7140/Tempus_Fugit_Minecraft/issues/20)
+        @see [Issue#28](https://github.com/WSUCEG-7140/Tempus_Fugit_Minecraft/issues/28)
```
Comment (bakursait2) [inline tempus_fugit_minecraft/world.py:231]: I believe this function was not created when we solve these two issues ([Issue#20](https://github.com/WSUCEG-7140/Tempus_Fugit_Minecraft/issues/20) , [Issue#28](https://github.com/WSUCEG-7140/Tempus_Fugit_Minecraft/issues/28)). instead I believe the related issues are:
([Issue#84](https://github.com/WSUCEG-7140/Tempus_Fugit_Minecraft/issues/84) , [Issue#86](https://github.com/WSUCEG-7140/Tempus_Fugit_Minecraft/issues/86))

**Review by bakursait2 — CHANGES_REQUESTED** [review]

File: tempus_fugit_minecraft/world.py (hunk 1 above)
Comment (criticalglitch) [reply tempus_fugit_minecraft/world.py:231]: You are correct that while this function wasn't created initially to sovle issues 20 and 28, the functions this replaces *was* meant to solve those two issues. This function is now a combination of `place_cloud_blocks`'s randomization of cloud color, and the block generative aspect of `generate_single_cloud`, both of which were originally in model.py.

I tagged this with #20 and #28 to show that this function still solves the problems those issues were created for and show how the code has evolved from when those issues were created to how they were initially fixed and now for this new evolution to solve the same issue.

**Comment by criticalglitch** [comment]
> I feel like before I can approve this, I need to understand why we need to split the model into model and world. As far as I can tell, when you move everything to world, you're pretty much just making a collection of static functions that the model will call. However most of these functions were also important to the model function. What exactly is the difference between the model and the world?

_As of right now_, there isn't much gained. We do get less code in the model, and a clearer understanding of how a world map generates since we now have nicely named functions that indicates what is being generated. However, this class will eventually gain the logic of managing the world, just not right now at first.

I actually tried at first to do a whole single split of the world and model (including all of the logic in the Model that contains it like `add_block`, `remove_block`, `show_block`, `hide_block`, etc...), however I ran into a lot more issues than I anticipated and decided to split that effort into three smaller PRs.

This is the first PR where the details of *how* to generate a world map have been split out. I'd like to split out world storage and interactability (aka the fields and functions that do not deal with sectors and block visibility) in the next PR of this series, then finally world rendering (the remaining bits of world logic) as the third PR to complete this issue.

This approach will allow me to work on this in smaller more manageable pieces and hopefully prevent a large regression like I had encountered with my first approach with this.

**Review by bakursait2 — APPROVED** [review]

**Review by xieWright — APPROVED** [review]
Sounds like a good explanation. Looks good then.

File: .github/workflows/python-app.yml [hunk 2]
```diff
@@ -21,7 +21,7 @@ jobs:
       # reference the matrixe python version here.
       - uses: actions/setup-python@v2
         with:
-          python-version: 3.9
+          python-version: 3.11
```
Comment (w1158mxs) [inline .github/workflows/python-app.yml:24]: I see Python upgraded here - we need to upgrade I next work

**Review by w1158mxs — APPROVED** [review]
What is the function that didn't work with the previous python version, why did you upgrade?

**Comment by criticalglitch** [comment]
> What is the function that didn't work with the previous python version, why did you upgrade?

I upgraded it because I wanted to leverage newer features in python like Type Aliases to provide some names for commonly used types of data. Like a `Position` being an alias for `tuple[int, int, int]`, and the world map being a type of `Map` which is an alias of `Dict[Position, Block]`. It helps provide a clearer meaning to what some of the things we were passing around or using are. 

Unfortunately, while type hints were added in python 3.5, the `TypeAlias` class in the `typing` module wasn't introduced until python 3.10. I upgraded CI to python 3.11 to match what my machine is using since I couldn't get poetry or the project to work with anything lower than python 3.11.

File: tests/test_world.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
+import random
+
+import pytest
+
+from tempus_fugit_minecraft.block import (BRICK, DARK_CLOUD, GRASS,
```
Comment (Lurk390) [inline tests/test_world.py:5]: Same suggestion as above, I recommended changing to reduce total line length:
```
from tempus_fugit_minecraft.block import (BRICK, DARK_CLOUD, GRASS, LIGHT_CLOUD, SAND, STONE, TREE_LEAVES, TREE_TRUNK, 
                                          Block)
```

**Review by Lurk390 — CHANGES_REQUESTED** [review]
Looks good overall, no issue with any logic or code itself. All my suggestions are documentation related and should be easy fixes.

File: tempus_fugit_minecraft/model.py [hunk 4]
```diff
@@ -78,52 +57,29 @@ def __init__(self) -> None:
...
 
-        self._initialize()
         self.background_noise = sound_list.wind_blowing
         self.background_noise.play_sound()
 
-    def _initialize(self, immediate=False) -> None:
+    def generate(self) -> None:
         """!
         @brief Initialize the world by placing all the blocks.
         @param immediate : True: draw block immediate; False: do not draw Block immediately. (default=False)
```
[the remark this answers is not in the document]
Comment (criticalglitch) [reply tempus_fugit_minecraft/model.py]: Done

File: tempus_fugit_minecraft/world.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
+import random
+import sys
+from typing import Dict, TypeAlias
+
+from tempus_fugit_minecraft.block import (BRICK, DARK_CLOUD, GRASS,
```
[the remark this answers is not in the document]
Comment (criticalglitch) [reply tempus_fugit_minecraft/world.py]: Done

File: tempus_fugit_minecraft/world.py [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+    x, y, z = position
+    x, y, z = (int(round(x)), int(round(y)), int(round(z)))
+    return x, y, z
+
+
+def sectorize(position: tuple) -> tuple:
+    """!
+    @brief Returns a tuple representing the sector for the given `position`.
+    @param position : tuple of len 3
+    @returns sector : tuple of len 3
```
[the remark this answers is not in the document]
Comment (criticalglitch) [reply tempus_fugit_minecraft/world.py]: Done

File: tempus_fugit_minecraft/world.py [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+
+if sys.version_info[0] >= 3:
+    xrange = range
+
+
+def normalize(position: tuple) -> Position:
+    """!
+    @brief Accepts `position` of arbitrary precision and returns the block containing that position.
+    @param position : tuple of len 3
+    @returns block_position : tuple of ints of len 3
```
[the remark this answers is not in the document]
Comment (criticalglitch) [reply tempus_fugit_minecraft/world.py]: Done

File: tempus_fugit_minecraft/world.py [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+class World:
+    """!
+    @brief A collection of functions to generate the structure 
+        of a 3D world
+    """
+    @staticmethod
+    def generate_base_layer() -> list[tuple[Block, Position]]:
+        """!
+        @brief Generate the base layer of the world
+        @returns a list of pairs of blocks and their positions
```
[the remark this answers is not in the document]
Comment (criticalglitch) [reply tempus_fugit_minecraft/world.py]: Done

File: tempus_fugit_minecraft/world.py [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+
+    x, y, z = normalize(position)
+    x, y, z = x // SECTOR_SIZE, y // SECTOR_SIZE, z // SECTOR_SIZE
+    return x, 0, z
+
+
+class World:
+    """!
+    @brief A collection of functions to generate the structure 
+        of a 3D world
```
[the remark this answers is not in the document]
Comment (criticalglitch) [reply tempus_fugit_minecraft/world.py]: Done

File: tempus_fugit_minecraft/world.py [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+
+    @staticmethod
+    def generate_hills(world_size = WORLD_SIZE, num_hills = int(WORLD_SIZE * 1.5)) -> list[list[tuple[Block, Position]]]:
+        """!
+        @brief this function generates a group of randomly positioned hills strewn around the world
+        @param world_size The world size (default: WORLD_SIZE)
+        @param num_hills The number of hills generated (default: 1.5x WORLD_SIZE)
+        @returns a list of lists of pairs of blocks and positions that represent a collection of hills
+        @see [Issue#86](https://github.com/WSUCEG-7140/Tempus_Fugit_Minecraft/issues/86)
+        """
```
[the remark this answers is not in the document]
Comment (criticalglitch) [reply tempus_fugit_minecraft/world.py]: Done

File: tempus_fugit_minecraft/world.py [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+            hills.append(hill)
+        return hills
+
+    @staticmethod
+    def generate_hill(center_x: int, center_z: int) -> list[tuple[Block, Position]]:
+        """!
+        @brief this function generates a single hill
+        @param center_x Represents the x coordinate center of the hill
+        @param center_z Represents the z coordinate center of the hill
+        @returns a list of pairs of blocks and positions that represent a hill
```
[the remark this answers is not in the document]
Comment (criticalglitch) [reply tempus_fugit_minecraft/world.py]: Done

File: tests/test_world.py (hunk 3 above)
Comment (criticalglitch) [reply tests/test_world.py]: Done

**Review by Lurk390 — APPROVED** [review]
Looks good, all of my suggestions were addressed.

**Pull request merged**